Global TSO sync failed due to the local tso allocator leader change

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What did you do?

See GitHub Actions test result: https://github.com/tikv/pd/runs/3043267045?check_suite_focus=true

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x23c5e19]

goroutine 13007 [running]:
github.com/tikv/pd/server.(*Server).SyncMaxTS(0xc00ecb0f20, 0x42f9e58, 0xc00fba6b70, 0xc00ddd9640, 0xc00ecb0f20, 0x3131e00, 0xc0001acd00)
	/home/runner/work/pd/pd/server/grpc_service.go:1323 +0xf79
github.com/pingcap/kvproto/pkg/pdpb._PD_SyncMaxTS_Handler.func1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0x0, 0x0, 0xc1cdcc, 0x187a00d)
	/home/runner/go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20210604082642-dda0a102bc6a/pkg/pdpb/pdpb.pb.go:6032 +0xa5
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0x12, 0xc013ebfea0, 0xc0012ff388, 0x2e445e0)
	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:31 +0x2d6
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0xc0115001e0, 0xc013ebfe50, 0xc00fba6a10, 0x5a891a8, 0x5a891a8, 0xc014321201)
	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0x105
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0x5a8b200, 0xe, 0x0, 0x0)
	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x1df
go.etcd.io/etcd/etcdserver/api/v3rpc.newUnaryInterceptor.func1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0xc0115001e0, 0xc013ebfe50, 0x0, 0x0, 0xc001d4d618, 0xc184bb)
	/home/runner/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:64 +0x1f1
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0x0, 0x0, 0x203004, 0x203004)
	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x1df
go.etcd.io/etcd/etcdserver/api/v3rpc.newLogUnaryInterceptor.func1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0xc0115001e0, 0xc013ebfe50, 0x0, 0x0, 0x0, 0x0)
	/home/runner/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:71 +0xd7
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x42f9e58, 0xc00fba6b70, 0x31edaa0, 0xc00ddd9640, 0xc0115001e0, 0xc0133797a0, 0xc001d4d918, 0xbb3498, 0x30, 0x3168bc0)
	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:39 +0x350
github.com/pingcap/kvproto/pkg/pdpb._PD_SyncMaxTS_Handler(0x328b900, 0xc00ecb0f20, 0x42f9e58, 0xc00fba6b70, 0xc00cb1a420, 0xc0149e3560, 0x42f9e58, 0xc00fba6b70, 0xc000d4c4a0, 0xe)
	/home/runner/go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20210604082642-dda0a102bc6a/pkg/pdpb/pdpb.pb.go:6034 +0x1f2
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0149ff380, 0x430ff78, 0xc0205cea80, 0xc015475400, 0xc014aad410, 0x4f37748, 0x0, 0x0, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024 +0xa23
google.golang.org/grpc.(*Server).handleStream(0xc0149ff380, 0x430ff78, 0xc0205cea80, 0xc015475400, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313 +0x13b8
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc02033eb80, 0xc0149ff380, 0x430ff78, 0xc0205cea80, 0xc015475400)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722 +0xe7
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:720 +0xb9
```

### What version of PD are you using (`pd-server -V`)?

#3850 (also the latest master 671b49f4bedaa0a233c71c6af7236609e588b856)

--- a/server/grpc_service.go
+++ b/server/grpc_service.go
@@ -1311,6 +1311,9 @@ func (s *Server) SyncMaxTS(ctx context.Context, request *pdpb.SyncMaxTSRequest) 
 			}
 			syncedDCs = append(syncedDCs, allocator.GetDCLocation())
 		}
+		if maxLocalTS == nil {
+			return nil, status.Errorf(codes.Unknown, "local tso allocator leaders have changed during the sync, should retry")
+		}
 		// Found a bigger or equal maxLocalTS, return it directly.
 		cmpResult := tsoutil.CompareTimestamp(maxLocalTS, request.GetMaxTs())
 		if cmpResult >= 0 {

--- a/server/grpc_service.go
+++ b/server/grpc_service.go
@@ -1277,6 +1277,9 @@ func (s *Server) incompatibleVersion(tag string) *pdpb.ResponseHeader {
 	})
 }
 
+// Only used for the TestLocalAllocatorLeaderChange.
+var mockLocalAllocatorLeaderChangeFlag = false
+
 // SyncMaxTS will check whether MaxTS is the biggest one among all Local TSOs this PD is holding when skipCheck is set,
 // and write it into all Local TSO Allocators then if it's indeed the biggest one.
 func (s *Server) SyncMaxTS(ctx context.Context, request *pdpb.SyncMaxTSRequest) (*pdpb.SyncMaxTSResponse, error) {
@@ -1311,8 +1314,20 @@ func (s *Server) SyncMaxTS(ctx context.Context, request *pdpb.SyncMaxTSRequest) 
 			}
 			syncedDCs = append(syncedDCs, allocator.GetDCLocation())
 		}
+
+		failpoint.Inject("mockLocalAllocatorLeaderChange", func() {
+			if !mockLocalAllocatorLeaderChangeFlag {
+				maxLocalTS = nil
+				request.MaxTs = nil
+				mockLocalAllocatorLeaderChangeFlag = true
+			}
+		})
+
 		if maxLocalTS == nil {
 			return nil, status.Errorf(codes.Unknown, "local tso allocator leaders have changed during the sync, should retry")
+		}
+		if request.GetMaxTs() == nil {
+			return nil, status.Errorf(codes.Unknown, "empty maxTS in the request, should retry")
 		}
 		// Found a bigger or equal maxLocalTS, return it directly.
 		cmpResult := tsoutil.CompareTimestamp(maxLocalTS, request.GetMaxTs())

--- a/server/tso/global_allocator.go
+++ b/server/tso/global_allocator.go
@@ -256,6 +256,10 @@ func (gta *GlobalTSOAllocator) precheckLogical(maxTSO *pdpb.Timestamp, suffixBit
 			globalTSOOverflowFlag = false
 		}
 	})
+	// Make sure the physical time is not empty again.
+	if maxTSO.GetPhysical() == 0 {
+		return false
+	}
 	// Check if the logical part will reach the overflow condition after being differenitated.
 	if differentiatedLogical := gta.timestampOracle.differentiateLogical(maxTSO.Logical, suffixBits); differentiatedLogical >= maxLogical {
 		log.Error("estimated logical part outside of max logical interval, please check ntp time",

--- a/server/tso/global_allocator.go
+++ b/server/tso/global_allocator.go
@@ -173,11 +173,11 @@ func (gta *GlobalTSOAllocator) GenerateTSO(count uint32) (pdpb.Timestamp, error)
 	defer cancel()
 	for i := 0; i < maxRetryCount; i++ {
 		var (
-			globalTSOResp          pdpb.Timestamp  = pdpb.Timestamp{}
-			estimatedMaxTSO        *pdpb.Timestamp = &pdpb.Timestamp{}
-			suffixBits             int             = gta.allocatorManager.GetSuffixBits()
-			shouldRetry, skipCheck bool
 			err                    error
+			shouldRetry, skipCheck bool
+			globalTSOResp          pdpb.Timestamp
+			estimatedMaxTSO        *pdpb.Timestamp
+			suffixBits             = gta.allocatorManager.GetSuffixBits()
 		)
 		// TODO: add a switch to control whether to enable the MaxTSO estimation.
 		// 1. Estimate a MaxTS among all Local TSO Allocator leaders according to the RTT.

--- a/server/tso/global_allocator.go
+++ b/server/tso/global_allocator.go
@@ -171,19 +171,20 @@ func (gta *GlobalTSOAllocator) GenerateTSO(count uint32) (pdpb.Timestamp, error)
 	// (whit synchronization with other Local TSO Allocators)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	var (
-		globalTSOResp          pdpb.Timestamp  = pdpb.Timestamp{}
-		estimatedMaxTSO        *pdpb.Timestamp = &pdpb.Timestamp{}
-		suffixBits             int             = gta.allocatorManager.GetSuffixBits()
-		shouldRetry, skipCheck bool
-		err                    error
-	)
 	for i := 0; i < maxRetryCount; i++ {
+		var (
+			globalTSOResp          pdpb.Timestamp  = pdpb.Timestamp{}
+			estimatedMaxTSO        *pdpb.Timestamp = &pdpb.Timestamp{}
+			suffixBits             int             = gta.allocatorManager.GetSuffixBits()
+			shouldRetry, skipCheck bool
+			err                    error
+		)
 		// TODO: add a switch to control whether to enable the MaxTSO estimation.
 		// 1. Estimate a MaxTS among all Local TSO Allocator leaders according to the RTT.
 		estimatedMaxTSO, shouldRetry, err = gta.estimateMaxTS(count, suffixBits)
 		if err != nil {
-			return pdpb.Timestamp{}, err
+			log.Error("global tso allocator estimates MaxTS failed", errs.ZapError(err))
+			continue
 		}
 		if shouldRetry {
 			time.Sleep(gta.timestampOracle.updatePhysicalInterval)
@@ -195,7 +196,8 @@ func (gta *GlobalTSOAllocator) GenerateTSO(count uint32) (pdpb.Timestamp, error)
 		// we need to validate it first before we write it into every Local TSO Allocator's memory.
 		globalTSOResp = *estimatedMaxTSO
 		if err = gta.SyncMaxTS(ctx, dcLocationMap, &globalTSOResp, skipCheck); err != nil {
-			return pdpb.Timestamp{}, err
+			log.Error("global tso allocator synchronizes MaxTS failed", errs.ZapError(err))
+			continue
 		}
 		// 3. If skipCheck is false and the maxTSO is bigger than estimatedMaxTSO,
 		// we need to redo the setting phase with the bigger one and skip the check safely.
@@ -218,15 +220,16 @@ func (gta *GlobalTSOAllocator) GenerateTSO(count uint32) (pdpb.Timestamp, error)
 		// 4. Persist MaxTS into memory, and etcd if needed
 		var currentGlobalTSO *pdpb.Timestamp
 		if currentGlobalTSO, err = gta.getCurrentTSO(); err != nil {
-			return pdpb.Timestamp{}, err
+			log.Error("global tso allocator gets the current global tso in memory failed", errs.ZapError(err))
+			continue
 		}
 		if tsoutil.CompareTimestamp(currentGlobalTSO, &globalTSOResp) < 0 {
 			tsoCounter.WithLabelValues("global_tso_persist", gta.timestampOracle.dcLocation).Inc()
 			// Update the Global TSO in memory
 			if err = gta.timestampOracle.resetUserTimestamp(gta.leadership, tsoutil.GenerateTS(&globalTSOResp), true); err != nil {
 				tsoCounter.WithLabelValues("global_tso_persist_err", gta.timestampOracle.dcLocation).Inc()
-				log.Error("update the global tso in memory failed", errs.ZapError(err))
-				return pdpb.Timestamp{}, err
+				log.Error("global tso allocator update the global tso in memory failed", errs.ZapError(err))
+				continue
 			}
 		}
 		// 5. Check leadership again before we returning the response.
@@ -240,7 +243,7 @@ func (gta *GlobalTSOAllocator) GenerateTSO(count uint32) (pdpb.Timestamp, error)
 		return globalTSOResp, nil
 	}
 	tsoCounter.WithLabelValues("exceeded_max_retry", gta.timestampOracle.dcLocation).Inc()
-	return globalTSOResp, errs.ErrGenerateTimestamp.FastGenByArgs("maximum number of retries exceeded")
+	return pdpb.Timestamp{}, errs.ErrGenerateTimestamp.FastGenByArgs("global tso allocator maximum number of retries exceeded")
 }
 
 // Only used for test

--- a/tests/server/tso/global_tso_test.go
+++ b/tests/server/tso/global_tso_test.go
@@ -559,3 +559,35 @@ func (s *testSynchronizedGlobalTSO) TestSynchronizedGlobalTSOOverflow(c *C) {
 	s.testGetTimestamp(ctx, c, cluster, tsoCount, tso.GlobalDCLocation)
 	failpoint.Disable("github.com/tikv/pd/server/tso/globalTSOOverflow")
 }
+
+func (s *testSynchronizedGlobalTSO) TestLocalAllocatorLeaderChange(c *C) {
+	c.Assert(failpoint.Enable("github.com/tikv/pd/server/mockLocalAllocatorLeaderChange", `return(true)`), IsNil)
+	defer failpoint.Disable("github.com/tikv/pd/server/mockLocalAllocatorLeaderChange")
+	dcLocationConfig := map[string]string{
+		"pd1": "dc-1",
+	}
+	dcLocationNum := len(dcLocationConfig)
+	cluster, err := tests.NewTestCluster(s.ctx, dcLocationNum, func(conf *config.Config, serverName string) {
+		conf.EnableLocalTSO = true
+		conf.Labels[config.ZoneLabel] = dcLocationConfig[serverName]
+	})
+	defer cluster.Destroy()
+	c.Assert(err, IsNil)
+
+	err = cluster.RunInitialServers()
+	c.Assert(err, IsNil)
+
+	cluster.WaitAllLeaders(c, dcLocationConfig)
+
+	s.leaderServer = cluster.GetServer(cluster.GetLeader())
+	c.Assert(s.leaderServer, NotNil)
+	s.dcClientMap[tso.GlobalDCLocation] = testutil.MustNewGrpcClient(c, s.leaderServer.GetAddr())
+	for _, dcLocation := range dcLocationConfig {
+		pdName := s.leaderServer.GetAllocatorLeader(dcLocation).GetName()
+		s.dcClientMap[dcLocation] = testutil.MustNewGrpcClient(c, cluster.GetServer(pdName).GetAddr())
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	s.testGetTimestamp(ctx, c, cluster, tsoCount, tso.GlobalDCLocation)
+}

